### PR TITLE
first attempt to implement matmul detection and optimization using th…

### PR DESCRIPTION
**Pull request by chelini**
…e matchers. Compared to the paper 'High-performace Generalized Tensor Operations: A compiler-Oriented Approach', the missing steps for detection are: check stride 1 for the induction variables i,j,k, check that the additional reads accesses that do not belong to matmul have stride 0 if the innermost loop is exchanged with any of the loops. Moreover, they also check that the last access is a write. Regrading optimization, the missing parts are: packing transformation and tile isolation. Moreover, I assume that the initial indexes are in the following order i-j-k with i outermost and manually set the tile sizes.

### CODE REVIEW

File: islutils/pet_wrapper.cc [hunk 1]
```diff
@@ -167,4 +167,8 @@ pet_stmt *Scop::stmt(isl::id id) const {
   return nullptr;
 }
 
+IslCopyRefWrapper<isl::schedule> Scop::schedule() {
```
Comment (ftynse) [inline islutils/pet_wrapper.cc:170]: This looks like code I have written.  How about keeping it in a separate commit?  This makes clear who wrote what and hence who's to blame when it breaks.  It also avoids painful rebase management when I attempt to merge one of the branches where I do have a separate commit adding exactly the same code (44d11724a50ef669bc9e0c13aaf5615894b8c535 
63cb7b8 )

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,7 @@ set(TEST_INPUTS
     strided_domain.c
     strided_domain_with_coefficients.c
     strided_domain_multi_dimensions.c
+    1mmWithOutInitStmt.c
```
Comment (ftynse) [inline test/CMakeLists.txt]: Nit: "without" is a single word in English, there is no sense in capitalizing "out"

File: test/inputs/1mmWithOutInitStmt.c [hunk 3]
```diff
@@ -0,0 +1,15 @@
+int tmp[1024][1024];
```
Comment (ftynse) [inline test/inputs/1mmWithOutInitStmt.c]: I would try with non-square matrices

File: test/test_transformer.cc [hunk 4]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
   ASSERT_TRUE(loop4pos > loop3pos);
   ASSERT_TRUE(stmtpos > loop4pos);
 }
+
+static isl::schedule_node tile(isl::schedule_node node,
+			       std::vector<int> tileSizes,
```
Comment (ftynse) [inline test/test_transformer.cc]: Please do _not_ mix tabs and spaces.  Generally, run clang-format with LLVM style on your code.

File: test/test_transformer.cc [hunk 5]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  			       int defaultSize = 32) {
```
Comment (ftynse) [inline test/test_transformer.cc]: I find this interface quite confusing, I would have made it fail on an assertion (it's a test after all!) if the size of `tileSizes` mismatches the number of band members.  Try writing documentation for this function to understand my complaint :)

File: test/test_transformer.cc [hunk 6]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  for(unsigned i = 0; i < dims; ++i) {
+    int tileSize = i < tileSizes.size() ? tileSizes[i] : defaultSize;
+    sizes = sizes.set_val(i, isl::val(node.get_ctx(), tileSize));
+  }
+  node = 
+    isl::manage(isl_schedule_node_band_tile(node.release(), sizes.release()));
+  return node;
+}
+
+static isl::multi_union_pw_aff permuteDims(isl::multi_union_pw_aff ps, int firstDim,
```
Comment (ftynse) [inline test/test_transformer.cc]: maybe call this `swapDims` it does not support arbitrary permutations

File: test/test_transformer.cc [hunk 7]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+					   int secondDim) {
+  auto scheduleFirstDim = ps.get_union_pw_aff(firstDim);
+  auto scheduleSecondDim = ps.get_union_pw_aff(secondDim);
+  ps = ps.set_union_pw_aff(secondDim, scheduleFirstDim);
+  ps = ps.set_union_pw_aff(firstDim, scheduleSecondDim);
+  return ps;
+}
+
+TEST(Transformer, matmul) {
```
Comment (ftynse) [inline test/test_transformer.cc]: Please capitalize the name of the test, it may also be helpful to make it more self-descriptive like `MatchMatmul`

File: test/test_transformer.cc [hunk 8]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  isl::schedule_node root = scop.schedule.get_root();
+ 
+  using namespace matchers;
+  isl::schedule_node node;
+  // clang-format off
+  auto matcher = band(
+    [&node] (isl::schedule_node n) {
+      if(isl_schedule_node_band_n_member(n.get()) < 3)
+        return false;
+      else { node = n; return true; }
```
Comment (ftynse) [inline test/test_transformer.cc]: Either use braces for both the "if" and "else" parts, or don't use them at all.  Also consider starting each statement from a new line of you have more than one and use braces.

File: test/test_transformer.cc [hunk 9]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+    },
+    leaf());
+  // clang-format on
+
+  bool matched = false;
+  if(ScheduleNodeMatcher::isMatching(matcher, root.child(0))) {
```
Comment (ftynse) [inline test/test_transformer.cc]: This control flow is useless.  Just do `auto matched = ScheduleNodeMatcher::isMatching(matcher, root.child(0));`

File: test/test_transformer.cc [hunk 10]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  auto _A = arrayPlaceholder();
+  auto _B = arrayPlaceholder();
+  auto _C = arrayPlaceholder();
+
+  auto psRead = allOf(access(_A, _i, _j), 
+		      access(_B, _i, _k),
+		      access(_C, _k, _j));
+  auto matchedReads = 
+    match(reads, psRead);
+  //std::cout << "matched Reads " << matchedReads.size() << std::endl;
```
Comment (ftynse) [inline test/test_transformer.cc]: Let's not commit commented-out code if possible.

File: test/test_transformer.cc [hunk 11]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  //std::cout << "matched Writes " << matchedWrites.size() << std::endl;
+
+  // tmp[j][i] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the checks. this is because we do not link
+  // read and write at the moment. Placeholder are _not_
+  // reused between different calls to allOf. We can overcome
+  // this inspecting the placeholder for the write and the read.
+  // They should be equal. 
+ 
+  matched = matched && (matchedWrites.size() == 1);
```
Comment (ftynse) [inline test/test_transformer.cc]: Why not just `ASSERT_EQ(matchedWrites.size(), 1u)` ?  It would fail the test early without executing anything else.

File: test/test_transformer.cc [hunk 12]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  matched = matched && (matchedReads.size() == 1);
```
Comment (ftynse) [inline test/test_transformer.cc]: The name `matchedReads` makes it slightly confusing.  It implies it contains the matched read relations.  I expect three of those, but you check there is only one.  (I know the variable contains _match descriptors_, indexed by placeholders.)  Maybe `readMatches` is a better name here.

File: test/test_transformer.cc [hunk 13]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+    match(reads, psRead);
+  //std::cout << "matched Reads " << matchedReads.size() << std::endl;
+  auto psWrite = allOf(access(_A, _ii, _jj));
+  auto matchedWrites =
+    match(writes, psWrite);
+  //std::cout << "matched Writes " << matchedWrites.size() << std::endl;
+
+  // tmp[j][i] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the checks. this is because we do not link
+  // read and write at the moment. Placeholder are _not_
```
Comment (ftynse) [inline test/test_transformer.cc]: Indeed.  Or figure out a way to reuse placeholders between different calls.  Or take a union of reads and writes and differentiate them by tags (may need a different pattern/candidate class than we already have).

File: test/test_transformer.cc [hunk 14]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  EXPECT_TRUE(matched);
+
+  // D[i][j] = alpha * A[i][k] * B[k][j] + tmp[i][j]
+  // pass the test. We may want to apply the same check
+  // as before also for the accessed array.
+
+  // step 1. (not implemented, just assumed)
+  // we assume the indexes are already in
+  // the "good" order. (i,j,k).
+  // Line 1285 to 1291 in ScheduleOptimizer.cpp
```
Comment (ftynse) [inline test/test_transformer.cc]: ScheduleOptimizer.cpp is not in this codebase, so I wouldn't base the documentation based on it.  Especially line numbers that are prone to changes without notice :)

Also, it should be easy to check the right order, you have access to `inputDimPos`...

File: test/test_transformer.cc [hunk 15]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  // step 2. create macro-kernel
+  // line 1294 in ScheduleOptimizer.cpp
+  // For the micro and macro kernels we assume
+  // given values for the tile size.
+  using namespace builders;
+  // tiling - Mc=96, Nc=2048, Kc=512
+  // set tile values manually
+  int dimOutNum = isl_schedule_node_band_n_member(node.get());
+  std::vector<int> tileSizes(dimOutNum);
+  tileSizes[dimOutNum - 3] = /*96*/32;
```
Comment (ftynse) [inline test/test_transformer.cc]: This contradicts the comment above.  And this is exactly the reason of me insisting on not committing commented-out code.

File: test/test_transformer.cc [hunk 16]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  tileSizes[dimOutNum - 2] = /*2048*/32;
+  tileSizes[dimOutNum - 1] = /*512*/32;
+  // tile node and get partial schedule 
+  isl::schedule_node tiledNode = tile(node, tileSizes);
+  auto tileSchedule = tiledNode.band_get_partial_schedule();
+  auto pointSchedule = tiledNode.child(0).band_get_partial_schedule();
+  // id for mark nodes.
```
Comment (ftynse) [inline test/test_transformer.cc]: mark nodes are useless, polly does them for annotation, but we don't need them in the test IMO

File: test/test_transformer.cc [hunk 17]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  // clang-format off
+  ScheduleNodeBuilder macroKernel = 
+    mark(idTile,
+      band(
+        [&node, &tileSchedule, &dimOutNum]() {
+          auto newPartialSchedule = permuteDims(tileSchedule, dimOutNum - 2, dimOutNum - 1);
+          newPartialSchedule = permuteDims(newPartialSchedule, dimOutNum - 3, dimOutNum - 1);
+          return newPartialSchedule;},
+        mark(idPoint,
+	  band(pointSchedule))));
```
Comment (ftynse) [inline test/test_transformer.cc]: one more indent here

File: test/test_transformer.cc [hunk 18]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  isl::id idPoint = isl::id::alloc(node.get_ctx(), markerPoint, nullptr);
+
+  // clang-format off
+  ScheduleNodeBuilder macroKernel = 
+    mark(idTile,
+      band(
+        [&node, &tileSchedule, &dimOutNum]() {
+          auto newPartialSchedule = permuteDims(tileSchedule, dimOutNum - 2, dimOutNum - 1);
+          newPartialSchedule = permuteDims(newPartialSchedule, dimOutNum - 3, dimOutNum - 1);
+          return newPartialSchedule;},
```
Comment (ftynse) [inline test/test_transformer.cc]: please put the closing brace on a new line

File: test/test_transformer.cc [hunk 19]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  if(matched) {
+    node = rebuild(node, macroKernel);
+    root = node.root();
+    //std::cout << root.to_str() << std::endl;
+  }
+
+  // match micro-kernel
+  auto matcherMicroKernel = [&]() {
+    using namespace matchers;
+    // clang-format off
```
Comment (ftynse) [inline test/test_transformer.cc]: there's no need to turn the formatter of if you don't have the tree formatting in can break :)

File: test/test_transformer.cc [hunk 20]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  // we assume the indexes are already in
+  // the "good" order. (i,j,k).
+  // Line 1285 to 1291 in ScheduleOptimizer.cpp
+
+  // step 2. create macro-kernel
+  // line 1294 in ScheduleOptimizer.cpp
+  // For the micro and macro kernels we assume
+  // given values for the tile size.
+  using namespace builders;
+  // tiling - Mc=96, Nc=2048, Kc=512
```
Comment (ftynse) [inline test/test_transformer.cc]: What are `Mc`, `Nc` and `Kc` ?   Your input code has fixed sizes...

File: test/test_transformer.cc [hunk 21]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+      return band(node,leaf());
+    // clang-format on
+  }();
+
+  matched = matched && 
+    ScheduleNodeMatcher::isMatching(matcherMicroKernel, root.child(0).child(0).child(0).child(0)); 
+  EXPECT_TRUE(matched);
+  
+  // create micro-kernel
+  // tiling Mr = 4 and Nr = 8 and 1.
```
Comment (ftynse) [inline test/test_transformer.cc]: What are `Mr` and `Nr` ?

File: test/test_transformer.cc [hunk 22]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  	      return descr;}))));
+  // clang-format on
+
+  if(matched) {
+    node = rebuild(node, microKernel);
+    root = node.root();
+    //std::cout << root.to_str() << std::endl;
+  }
+/*
+  // isolate full-tile from partial tile
```
Comment (ftynse) [inline test/test_transformer.cc]: let's node commit commented-out code.

File: test/test_transformer.cc [hunk 23]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  }();
+
+  matched = matched &&
+    ScheduleNodeMatcher::isMatching(matcherMicroKernelAfterTiling, 
+      root.child(0).child(0).child(0).child(0).child(0).child(0).child(0));
+  EXPECT_TRUE(matched);
+*/    
+  std::cout << petScop.codegen() << std::endl;
+  petScop.schedule() = root.get_schedule();
+  std::cout << petScop.codegen() << std::endl;
```
Comment (ftynse) [inline test/test_transformer.cc]: How about you check for _some_ pattern in the code instead of just dumping it.  Like the new loops are indeed produced or something?

File: test/test_transformer.cc [hunk 24]
```diff
@@ -518,3 +518,211 @@ TEST(Transformer, Codegen) {
...
+  using namespace builders;
+  // tiling - Mc=96, Nc=2048, Kc=512
+  // set tile values manually
+  int dimOutNum = isl_schedule_node_band_n_member(node.get());
+  std::vector<int> tileSizes(dimOutNum);
+  tileSizes[dimOutNum - 3] = /*96*/32;
+  tileSizes[dimOutNum - 2] = /*2048*/32;
+  tileSizes[dimOutNum - 1] = /*512*/32;
+  // tile node and get partial schedule 
+  isl::schedule_node tiledNode = tile(node, tileSizes);
```
Comment (ftynse) [inline test/test_transformer.cc]: I don't quite get this.  You transform the node using isl's function, then essentially drop the transformed node and transform the original node _again_ with builders.  What's the point?  Maybe just extract the "tile" function that computes a pair of schedules (tiles, points) given the original schedule and sizes?

**Review by ftynse — COMMENTED** [review]
I'm a little nitpicky on this one, but seems to go in the right direction.  I wish we did it earlier.

File: test/test_transformer.cc (hunk 5 above)
Comment (chelini) [reply test/test_transformer.cc]: Ok, I think I got your point. In this case, if I want to use assert in the method I should you cassert or the one provided by googletest? It is ok to mix both?

File: test/test_transformer.cc (hunk 4 above)
Comment (chelini) [reply test/test_transformer.cc]: Yeah sorry, I forgot to run clang-format.

File: islutils/pet_wrapper.cc (hunk 1 above)
Comment (chelini) [reply islutils/pet_wrapper.cc:170]: Yes, it is your code. Can we integrate it into the master?

File: test/test_transformer.cc (hunk 14 above)
Comment (chelini) [reply test/test_transformer.cc]: Got your point I will do it.

File: test/test_transformer.cc (hunk 16 above)
Comment (chelini) [reply test/test_transformer.cc]: Ok I can remove them I think with mark not it is easy to check what's going on while you are transforming.

File: test/test_transformer.cc (hunk 17 above)
Comment (chelini) [reply test/test_transformer.cc]: mm, I saw it differently maybe because I am mixing tab with space.

File: test/test_transformer.cc (hunk 24 above)
Comment (chelini) [reply test/test_transformer.cc]: Yes, it is a bit silly. I just want to get the new schedule for the tile and the point tile. To do so I transform the node and get the new partial schedules and re-build it. Do you have suggestions to improve it? What do you mean by "extracting the tile function"?

File: test/test_transformer.cc (hunk 20 above)
Comment (chelini) [reply test/test_transformer.cc]: These are tile size values. The values are obtained using an analytical model that makes use of TargetTransform pass in LLVM. I basically got the values for my machine and check that I am generating the same code as Polly does.

File: islutils/pet_wrapper.cc (hunk 1 above)
Comment (ftynse) [reply islutils/pet_wrapper.cc:170]: Just don't meld it into your commit,  [cherry-pick](https://git-scm.com/docs/git-cherry-pick) the original commits to your branch if you need them.

File: test/test_transformer.cc (hunk 5 above)
Comment (ftynse) [reply test/test_transformer.cc]: You cannot use those from googletest unless you are inside a function declared as test, so just use the C one.  Or have a manual check and call `ISLUTILS_DIE` from `die.h`.

File: test/test_transformer.cc (hunk 16 above)
Comment (ftynse) [reply test/test_transformer.cc]: Keep them if you want but use them in checks then :)

File: test/test_transformer.cc (hunk 20 above)
Comment (ftynse) [reply test/test_transformer.cc]: Ok, just explain what `Mc` and `Nc` etc. mean and where they should go

File: test/test_transformer.cc (hunk 24 above)
Comment (ftynse) [reply test/test_transformer.cc]: You don't need to transform the node, you just need the two `multi_union_pw_aff`s for the parent and the child node.  There may be a function that does that in isl, or it's quite easy to implement one (divide by multi-val, or take modulo by multival).

File: test/test_transformer.cc (hunk 4 above)
Comment (pfaffe) [reply test/test_transformer.cc]: You can make clang-format into a git precommit hook, either to auto-format your commit or to reject it so that you can format it.

**Review by ftynse — COMMENTED** [review]
Thanks.  Please squash all "fix" commits into the first one and split its description into a short part (76 characters) and a longer part (everything else separated by a newline from the first 76 characters); and it will be ready to go.

**Comment by chelini** [comment]
I have squashed the commits and push back. Please let me know if it looks good.

**Comment by ftynse** [comment]
Rebased and pushed to master as 6761257 

I forgot that I asked you to separate out the pre-existing code before asking you to squash everything, so I did that myself by merging #9 and rebasing.  

Generally, please consider making your git commits as unitary logical changes, not just snapshots of your working directory at some given time.  This helps tracking down problematic changes or reverting them if necessary.  That's why I ask to squash in review changes, and separate out changes that were made elsewhere.

**Comment by chelini** [comment]
Thanks Alex.
